lib/mini_mime.rb

```diff
@@ -102,6 +102,16 @@ def initialize(path, sort_order)
         @file_length = File.size(@path)
         @rows = @file_length / @row_length
 
+        # Make sure the DB file is seekable, use a StringIO wrapper if it's not
```

Comment:
This whole dance of checking makes me uneasy... I would be comfortable with a "reaction" to an error JRuby raises when you do a seek but it gives us nothing. I prefer not to own this heuristic.

Perhaps a simple:

```
MiniMime.preload_dbs = true
```

Then 

```
if MiniMime.preload_dbs
   @file = StringIO.new(File.read(@path))
end
```

With some readme data. Very uneasy with the hack of ... seek around file and check that the computer did what you asked it... as a general solution. @headius - it should be exceptioning out if you tell it to seek and it does not seek, this is hiding bugs in JRuby and making them "normal".

Reply:
> it should be exceptioning out if you tell it to seek and it does not seek

[I completely agree](https://github.com/discourse/mini_mime/issues/37#issuecomment-901488891). However, if we want this code to be compatible with current and past JRuby behavior, we'd still need this "hack". At least the way I wrote it, it should continue working even if JRuby behavior is later changed to raise a reasonable exception in this situation.

I suppose a comment explaining why this "dance" is needed might be worth including.

Reply:
We do not do anything special for `seek`. It will either use JVM functions to do this (SeekableChannel.position) or a native `tell` call.

It does appear to error for an `IO.pipe`, for example, but I'm unsure if it is the right error:

```
$ jruby -e 'IO.pipe[0].seek(0)'
Errno::EPIPE: Broken pipe - No message available
    seek at org/jruby/RubyIO.java:1817
  <main> at -e:1
```

Also for stdin, with a similar error:

```
$ jruby -e '$stdin.seek(0)'
Errno::ESPIPE: Illegal seek - <STDIN>
    seek at org/jruby/RubyIO.java:1817
  <main> at -e:1
```

I am a bit confused which resources are not seekable but not raising an error. Just classloader resources? Some other file type for the DB? I think we can extend this error raising to those other resources types if someone can clarify which ones are failing to raise an error here.

Reply:
I commented on my PR as well, but to keep this up-to-date...

I agree that classloader resources, accessible only as InputStream objects that don't support seeking, should raise an error when you attempt to seek on them. If we can be sure this won't affect rack, it is an easy change (and we could narrow my PR to just the resource types in question). Until we can be sure of that, however, we are potentially trading one "broken" behavior for another "broken" behavior.

The simplest fix would be to make all unseekable resources in JRuby raise an error, but there are many resource types in JRuby that would **never be encountered in CRuby** such as IO resources from in-memory resources that can only be read forward. It is an arguable grey area what is appropriate to do for those.